crates/server/src/store/in_memory/authority.rs

```diff
@@ -806,17 +806,24 @@ impl InnerInMemory {
                     entry.insert((HashSet::from([key.record_type]), true));
                 }
             }
+        }
 
-            // For every domain name between the current name and the origin, add it to
-            // `record_types` without any record types. This covers all the empty non-terminals
-            // that must have an NSEC3 record as well.
-            let mut name = key.name.base_name();
+        if opt_out {
+            // Delete owner names that have unsigned delegations.
+            let ns_only = HashSet::from([RecordType::NS]);
+            record_types.retain(|_name, (types, _exists)| types != &ns_only);
+        }
 
-            for _ in origin.num_labels()..name.num_labels() {
-                if let Entry::Vacant(entry) = record_types.entry(name.clone()) {
+        // For every domain name between the current name and the origin, add it to `record_types`
+        // without any record types. This covers all the empty non-terminals that must have an NSEC3
+        // record as well.
+        for name in record_types.keys().cloned().collect::<Vec<_>>() {
+            let mut parent = name.base_name();
+            while parent.num_labels() > origin.num_labels() {
+                if let Entry::Vacant(entry) = record_types.entry(parent.clone()) {
```

Comment:
Nit: I think you could use `Entry::or_insert_with((HashSet::new(), false))` to make this slightly more concise.